### PR TITLE
feat(chunkers): add padding as a swept chunking dimension (#45)

### PR DESCRIPTION
**Pull request by entzyeung**
Implements #45.

## What
Adds `padding` as a sweep dimension. It enforces a **minimum chunk character length**: any chunk shorter than `padding` is merged forward into the following chunk(s) until it reaches the threshold, avoiding low-signal fragment chunks (e.g. a trailing partial sentence) that embed poorly.

- Applied uniformly as post-processing after every chunker, so `padding` means the same thing across fixed/recursive/token/sentence/semantic.
- `padding=0` (the default) is a no-op — **existing sweeps are completely unchanged**.

## Plumbing
Threads `padding` through `ChunkParams.paddings` → `expand_sweep` Cartesian product → `RunParams` → `RunStatus` → the `chunk_text` dispatcher → orchestrator persistence (run status + chunk docs). Documents `paddings:` in `configs/example-mongodb-local.yaml`.

## Tests
`tests/test_padding.py` covers `_apply_padding` (no-op cases, forward-merge to threshold, trailing-remainder folding, already-large chunks) and that `paddings` expands as a sweep dimension. Local gates green (ruff + mypy + full pytest, 34 passed).

## Open question (re: the issue's design discussion)
The "minimum chunk size" interpretation here is one reasonable meaning of padding — happy to adjust to whatever semantics you'd prefer (e.g. fixed boundary context, or pad-to-token-multiple). If you'd rather a different definition, only the `_apply_padding` helper changes; the config/plumbing stays.

**Frontend display (padding badge + sweep-dimensions panel) is intentionally deferred** until the semantics are confirmed, to avoid building speculative UI.

### CODE REVIEW

File: server/core/chunkers/__init__.py [hunk 1]
```diff
@@ -4,17 +4,54 @@
...
+    Padding enforces a minimum chunk character length: any chunk shorter than
+    `padding` is merged with the following chunk(s) until it reaches the
+    threshold. This avoids low-signal fragment chunks (e.g. a trailing partial
+    sentence) that embed poorly. A final undersized chunk with nothing after it
+    is left as-is. `padding <= 0` is a no-op and returns chunks unchanged.
+
+    Method-agnostic by design: applied uniformly after every chunker so padding
+    means the same thing across fixed/recursive/token/sentence/semantic.
+    """
+    if padding <= 0 or len(chunks) <= 1:
```
Comment (neomatrix369) [inline server/core/chunkers/__init__.py:19]: **SUGGESTION** — The `len(chunks) <= 1` early-return correctly handles single-chunk and empty-list inputs, but the empty-list case isn't explicitly tested. I ran it manually:

```python
_apply_padding([], 10)  # → []  (correct — nothing to merge)
```

Would you consider adding `test_empty_list_unchanged`? It doubles as documentation that the contract is intentional and guards against a future refactor accidentally breaking it.

File: server/models/config.py [hunk 2]
```diff
@@ -13,6 +13,9 @@
 class ChunkParams(BaseModel):
     chunk_sizes: list[int] = Field(default=[512])
     overlaps: list[int] = Field(default=[50])
+    # Minimum chunk character length. Undersized chunks are merged forward into
```
Comment (neomatrix369) [inline server/models/config.py:16]: **SUGGESTION** — What do you think about adding a Pydantic `model_validator` that warns when any `padding` value exceeds the minimum `chunk_size`?

I tested the edge case locally:
```python
_apply_padding(["x" * 200] * 5, padding=1000)
# → 1 chunk of 1004 chars (all 5 chunks merged into one blob)
```

If someone writes `paddings: [300]` with `chunk_sizes: [256]`, every run silently produces a single massive chunk — which defeats chunking entirely. The behavior is correct per the algorithm, but is almost certainly a config mistake. A light-touch validator at config-load time would surface the problem before a sweep runs:

```python
@model_validator(mode="after")
def warn_padding_exceeds_chunk_size(self) -> "ChunkParams":
    min_size = min(self.chunk_sizes)
    bad = [p for p in self.paddings if p > min_size]
    if bad:
        import warnings
        warnings.warn(
            f"paddings {bad} exceed min chunk_size {min_size}: "
            "chunks may all merge into a single blob",
            stacklevel=2,
        )
    return self
```

Happy to defer if you'd prefer to keep validators out of `ChunkParams` for now — could also be a follow-up ticket.

File: server/models/config.py [hunk 3]
```diff
@@ -214,11 +219,12 @@ def expand_sweep(config: ExperimentConfig) -> list[RunParams]:
         config.chunking.methods,
```
Comment (neomatrix369) [inline server/models/config.py:219]: **NIT** — The `paddings or [0]` fallback is a nice defensive touch. Since the Pydantic field already defaults to `[0]`, `paddings` will never be falsy in practice. What do you think about adding a brief inline comment to make the intent clear to future readers, e.g.:

```python
config.chunking.params.paddings or [0],  # fallback: YAML [] would skip padding dimension
```

Without it, the `or [0]` looks like redundant dead code.

**Review by neomatrix369 — COMMENTED** [review]
## Smoke test results — PR #48 `feat(chunkers): add padding as a swept chunking dimension`

Checked out `feat/padding-sweep-dimension` and ran the full quality gate suite locally.

| Gate | Result |
|---|---|
| `ruff check .` | ✅ 0 errors |
| `mypy server/ cli/` | ✅ 0 errors, 52 files |
| `pytest` (full suite) | ✅ **86/86 passed** (8 new, 78 existing) |

**Backward compatibility confirmed**: existing tests are completely unaffected — `padding=0` is a true no-op.

Inline comments below flag one edge case worth discussing before merge, plus a few smaller observations.

File: tests/test_padding.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+
+def test_single_chunk_unchanged() -> None:
+    assert _apply_padding(["short"], 100) == ["short"]
+
+
+def test_undersized_chunks_merge_forward_to_threshold() -> None:
+    # Each fragment is 3 chars; threshold 7 forces pairs to merge (3+1+3=7).
+    chunks = ["aaa", "bbb", "ccc", "ddd"]
+    assert _apply_padding(chunks, 7) == ["aaa bbb", "ccc ddd"]
```
Comment (neomatrix369) [inline tests/test_padding.py:41]: **NIT** — The comment `# Each fragment is 3 chars; threshold 7 forces pairs to merge (3+1+3=7)` is correct, but the `+1+` (space from the join) is easy to skim past and misread as `3+3=6`. What do you think about being explicit: `# "aaa" + " " + "bbb" = 7 chars → meets threshold`? Small thing, but this is the test that explains the join mechanic to future readers.

File: server/core/chunkers/__init__.py (hunk 1 above)
Comment (neomatrix369) [reply server/core/chunkers/__init__.py:19]: Addressed in [`951dfcf`](https://github.com/neomatrix369/rag-params-finder/commit/951dfcf) (merged via [#60](https://github.com/neomatrix369/rag-params-finder/pull/60)):

Added `test_empty_list_unchanged` in `tests/test_padding.py` asserting `_apply_padding([], 10) == []`.

File: server/models/config.py (hunk 2 above)
Comment (neomatrix369) [reply server/models/config.py:16]: Addressed in [`951dfcf`](https://github.com/neomatrix369/rag-params-finder/commit/951dfcf) (merged via [#60](https://github.com/neomatrix369/rag-params-finder/pull/60)):

Added `ChunkParams.warn_padding_exceeds_chunk_size` (`@model_validator(mode="after")`) plus `test_padding_exceeds_chunk_size_emits_warning`.

File: server/models/config.py (hunk 3 above)
Comment (neomatrix369) [reply server/models/config.py:219]: Addressed in [`951dfcf`](https://github.com/neomatrix369/rag-params-finder/commit/951dfcf) (merged via [#60](https://github.com/neomatrix369/rag-params-finder/pull/60)):

Added inline comment on `expand_sweep`: `# fallback: YAML [] would skip padding dimension`.

File: tests/test_padding.py (hunk 4 above)
Comment (neomatrix369) [reply tests/test_padding.py:41]: Addressed in [`951dfcf`](https://github.com/neomatrix369/rag-params-finder/commit/951dfcf) (merged via [#60](https://github.com/neomatrix369/rag-params-finder/pull/60)):

Updated test comment to `# "aaa" + " " + "bbb" = 7 chars → meets threshold.`.
